optuna/storages/_rdb/alembic/versions/v3.0.0.c.py

```diff
@@ -61,31 +61,36 @@ def intermediate_value_to_stored_repr(
 
 def upgrade():
     bind = op.get_bind()
+    inspector = sa.inspect(bind)
+    column_names = [c["name"] for c in inspector.get_columns("trial_intermediate_values")]
 
     sa.Enum(IntermediateValueModel.TrialIntermediateValueType).create(bind, checkfirst=True)
 
     # MySQL and PostgreSQL supports DEFAULT clause like 'ALTER TABLE <tbl_name>
     # ADD COLUMN <col_name> ... DEFAULT "FINITE_OR_NAN"', but seemingly Alembic
     # does not support such a SQL statement. So first add a column with schema-level
     # default value setting, then remove it by `batch_op.alter_column()`.
-    with op.batch_alter_table("trial_intermediate_values") as batch_op:
-        batch_op.add_column(
-            sa.Column(
+    if "intermediate_value_type" not in column_names:
+        with op.batch_alter_table("trial_intermediate_values") as batch_op:
+            batch_op.add_column(
+                sa.Column(
+                    "intermediate_value_type",
+                    sa.Enum(
+                        "FINITE", "INF_POS", "INF_NEG", "NAN", name="trialintermediatevaluetype"
+                    ),
+                    nullable=False,
+                    server_default="FINITE",
+                ),
+            )
+        with op.batch_alter_table("trial_intermediate_values") as batch_op:
+            batch_op.alter_column(
                 "intermediate_value_type",
-                sa.Enum("FINITE", "INF_POS", "INF_NEG", "NAN", name="trialintermediatevaluetype"),
-                nullable=False,
-                server_default="FINITE",
-            ),
-        )
-    with op.batch_alter_table("trial_intermediate_values") as batch_op:
-        batch_op.alter_column(
-            "intermediate_value_type",
-            existing_type=sa.Enum(
-                "FINITE", "INF_POS", "INF_NEG", "NAN", name="trialintermediatevaluetype"
-            ),
-            existing_nullable=False,
-            server_default=None,
-        )
+                existing_type=sa.Enum(
+                    "FINITE", "INF_POS", "INF_NEG", "NAN", name="trialintermediatevaluetype"
+                ),
+                existing_nullable=False,
+                server_default=None,
+            )
```

Comment:
I found that the final schema had unexpected default value if an error occurred between `batch_op.add_column` and `batch_op.alter_column`. For example, `intermediate_value` had `DEFAULT 'FINITE'` in the case of SQLite3.

```sql
CREATE TABLE IF NOT EXISTS "trial_intermediate_values" (
        trial_intermediate_value_id INTEGER NOT NULL, 
        trial_id INTEGER NOT NULL, 
        step INTEGER NOT NULL, 
        intermediate_value FLOAT, intermediate_value_type VARCHAR(7) DEFAULT 'FINITE' NOT NULL, 
        PRIMARY KEY (trial_intermediate_value_id), 
        UNIQUE (trial_id, step), 
        FOREIGN KEY(trial_id) REFERENCES trials (trial_id)
);
```


I don't think it occurs frequently, but how about applying `batch_op.alter_column` even if `trial_intermediate_values` exist?

Reply:
Thank you for your review! It sounds reasonable to add the change to `v3.0.0.c.py`.
Does the same apply to `v3.0.0.d.py`?

Although there is a problem that a temporary table remains when using SQLite3 as discussed in https://github.com/sqlalchemy/alembic/discussions/1092, it may be useful for PostgreSQL and MySQL users.

Reply:
> Does the same apply to v3.0.0.d.py?

I checked if the proposal worked as follows:

1. Newly created tables with this PR
2. Migrated table from v2.10.0 without errors.
3. Migrated table from v2.10.0 with an error after [`add_column` in v3.0.0.c.py](https://github.com/optuna/optuna/blob/d42ab5c5c87dcd0a680f04361076976dd499290c/optuna/storages/_rdb/alembic/versions/v3.0.0.c.py#L72-L79)
4. Migrated table from v2.10.0 with an error after [`alter_column` in v3.0.0.c.py](https://github.com/optuna/optuna/blob/d42ab5c5c87dcd0a680f04361076976dd499290c/optuna/storages/_rdb/alembic/versions/v3.0.0.c.py#L80-L88)
5. Migrated table from v2.10.0 with an error during [the data migration in v3.0.0.c.py](https://github.com/optuna/optuna/blob/d42ab5c5c87dcd0a680f04361076976dd499290c/optuna/storages/_rdb/alembic/versions/v3.0.0.c.py#L119-L120)
6. Migrated table from v2.10.0 with an error after [`add_column` in v3.0.0.d.py](https://github.com/optuna/optuna/blob/d42ab5c5c87dcd0a680f04361076976dd499290c/optuna/storages/_rdb/alembic/versions/v3.0.0.d.py#L81-L89)
7. Migrated table from v2.10.0 with an error after [`alter_column` in v3.0.0.d.py](https://github.com/optuna/optuna/blob/d42ab5c5c87dcd0a680f04361076976dd499290c/optuna/storages/_rdb/alembic/versions/v3.0.0.d.py#L90-L101)
8. Migrated table from v2.10.0 with an error during [the data migration in v3.0.0.d.py](https://github.com/optuna/optuna/blob/d42ab5c5c87dcd0a680f04361076976dd499290c/optuna/storages/_rdb/alembic/versions/v3.0.0.d.py#L127-L128)

In all cases, the migrated tables (2 to 8) had the same schema as the newly created table (1). So, I don't think this proposal has major side-effect and I think it seems reasonable to apply to both v3.0.0.c.py and v3.0.0.d.py.
(As you mentioned, I needed to remove temporary tables manually in the case of (4), (5), (7) and (8).)

Reply:
Thank you for your thorough investigation! I pushed 006c02be4.